File: rust-sdk/client/src/pda.rs
use crate::{DEFAULT_PUSH_ORACLE_PROGRAM_ID, TUNA_ID};
use solana_pubkey::Pubkey;

pub fn get_tuna_config_address() -> (Pubkey, u8) {
    let seeds = &[b"tuna_config".as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_market_address(pool: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"market", pool.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_vault_address(token_mint: &Pubkey, market: Option<&Pubkey>) -> (Pubkey, u8) {
    if market.is_some() {
        let seeds = &[b"vault", token_mint.as_ref(), market.unwrap().as_ref()];
        Pubkey::find_program_address(seeds, &TUNA_ID)
    } else {
        let seeds = &[b"vault", token_mint.as_ref()];
        Pubkey::find_program_address(seeds, &TUNA_ID)
    }
}

pub fn get_lending_position_address(authority: &Pubkey, mint_or_vault: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"lending_position", authority.as_ref(), mint_or_vault.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_tuna_liquidity_position_address(position_mint: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"tuna_position", position_mint.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_tuna_spot_position_address(authority: &Pubkey, pool: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"tuna_spot_position", authority.as_ref(), pool.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_tuna_price_update_address(mint: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"tuna_price_update", mint.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

pub fn get_pyth_price_update_account_address(shard_id: u16, feed_id: [u8; 32]) -> (Pubkey, u8) {
    Pubkey::find_program_address(&[&shard_id.to_le_bytes(), feed_id.as_ref()], &DEFAULT_PUSH_ORACLE_PROGRAM_ID)
}

pub fn get_referral_address(authority: &Pubkey) -> (Pubkey, u8) {
    let seeds = &[b"referral", authority.as_ref()];
    Pubkey::find_program_address(seeds, &TUNA_ID)
}

File: ts-sdk/client/src/pda.ts
import { Address, getU32Encoder, ProgramDerivedAddress } from "@solana/kit";
import { getAddressEncoder, getProgramDerivedAddress } from "@solana/kit";

import { DEFAULT_PUSH_ORACLE_PROGRAM_ID } from "./consts.ts";
import { TUNA_PROGRAM_ADDRESS } from "./generated";

export async function getTunaConfigAddress(): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["tuna_config"],
  });
}

export async function getMarketAddress(pool: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["market", getAddressEncoder().encode(pool)],
  });
}

export async function getLendingVaultAddress(mint: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["vault", getAddressEncoder().encode(mint)],
  });
}

export async function getLendingVaultV2Address(mint: Address, id: number): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["vault", getAddressEncoder().encode(mint), getU32Encoder().encode(id)],
  });
}

export async function getLendingPositionAddress(
  authority: Address,
  mintOrVault: Address,
): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["lending_position", getAddressEncoder().encode(authority), getAddressEncoder().encode(mintOrVault)],
  });
}

export async function getTunaLpPositionAddress(positionMint: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["tuna_position", getAddressEncoder().encode(positionMint)],
  });
}

export async function getTunaSpotPositionAddress(authority: Address, pool: Address): Promise<ProgramDerivedAddress> {
  return await getProgramDerivedAddress({
    programAddress: TUNA_PROGRAM_ADDRESS,
    seeds: ["tuna_spot_position", getAddressEncoder().encode(authority), getAddressEncoder().encode(pool)],
  });
}

export async function getPythPriceUpdateAccountAddress(
  shardId: number,
  priceFeedId: Buffer | string,
): Promise<ProgramDerivedAddress> {
  if (typeof priceFeedId == "string") {
    if (priceFeedId.startsWith("0x")) {
      priceFeedId = Buffer.from(priceFeedId.slice(2), "hex");
    } else {
      priceFeedId = Buffer.from(priceFeedId, "hex");
    }
  }

  if (priceFeedId.length != 32) {
    throw new Error("Feed ID should be 32 bytes long");
  }
  const shardBuffer = Buffer.alloc(2);
  shardBuffer.writeUint16LE(shardId, 0);

  return await getProgramDerivedAddress({
    programAddress: DEFAULT_PUSH_ORACLE_PROGRAM_ID,
    seeds: [shardBuffer, priceFeedId],
  });
}
